rectools/visuals/metrics_app.py

```diff
@@ -85,125 +97,267 @@ def construct(
+        splits_set = set(splits)
+        if splits_set > 1:
+            raise ValueError(f"All models must have the same splits. But now they are different: {splits_set}")
+
+    @staticmethod
+    def _validate_models_metrics_names(models_metrics: pd.DataFrame) -> None:
+        # Validate that all Models names are unique
+        if Columns.Split in models_metrics.columns:
+            models_names_comb = models_metrics[Columns.Model] + models_metrics[Columns.Split].astype(str)
+            if models_names_comb.nunique() != len(models_names_comb):
+                raise ValueError("Each `Model` value in the `metrics_data` DataFrame must be unique")
+        else:
+            if models_metrics[Columns.Model].nunique() != len(models_metrics):
+                raise ValueError("Each `Model` value in the `metrics_data` DataFrame must be unique")
+
+    @staticmethod
+    def _validate_models_metadata(models_metadata: pd.DataFrame) -> None:
+        if Columns.Model not in models_metadata.columns:
+            raise KeyError("Missing `Model` column in `models_metadata` DataFrame")
+        if models_metadata[Columns.Model].isnull().any():
+            raise ValueError("Found NaN values in `Model` column")
+        if models_metadata[Columns.Model].nunique() != len(models_metadata):
+            raise ValueError("`Model` values of `models_metadata` should be unique`")
 
     @lru_cache
-    def _make_chart_data(self, fold_number: int) -> pd.DataFrame:  # pragma: no cover
-        return self.models_metrics[self.models_metrics[Columns.Split] == fold_number].drop(columns=Columns.Split)
+    def _make_chart_data_fold(self, fold_number: int) -> pd.DataFrame:
+        return self.data[self.data[Columns.Split] == fold_number].reset_index(drop=True)
 
     @lru_cache
     def _make_chart_data_avg(self) -> pd.DataFrame:
-        return self.models_metrics.drop(columns=Columns.Split).groupby(Columns.Model, sort=False).mean().reset_index()
+        metric_data_columns = [Columns.Model] + self.metric_names
+        meta_data_columns = [Columns.Model] + self.meta_names
+        # separate metric data because meta data could contain strings
+        metrics_data = self.data[metric_data_columns].groupby([Columns.Model], sort=False).mean().reset_index()
+        meta_data = self.data[meta_data_columns].drop_duplicates()
+        return metrics_data.merge(meta_data, on=Columns.Model, how="left").reset_index(drop=True)
+
+    @staticmethod
+    @lru_cache
+    def _split_to_meta_and_model(raw_string: str, splitter: str = ", ") -> tp.Tuple[str, str]:
+        splitted_row = raw_string.split(splitter, 1)
+        if len(splitted_row) > 1:
+            meta_value, model_name = splitted_row
+            return meta_value, model_name
+        return "", raw_string
+
+    def _create_chart_figure(
+        self,
+        data: pd.DataFrame,
+        metric_x: str,
+        metric_y: str,
+        color: str,
```

Comment:
Let's rename it to `color_col` please, now it's really confusing 

Same for metric_x and metric_y

Reply:
Done